stellargraph/data/unsupervised_sampler.py

```diff
@@ -121,69 +121,45 @@ def generator(self, batch_size):
-                # Get 1 walk at a time. For now its assumed that its a uniform random walker
-                walk = self.walker.run(
-                    nodes=[node],  # root nodes
-                    length=self.length,  # maximum length of a random walk
-                    n=1,  # number of random walks per root node
-                )
+        walks = self.walker.run(nodes=self.nodes, length=self.length, n=1, seed=0)
+
+        # first item in each walk is the target/head node
+        targets = [walk[0] for walk in walks]
+
+        positive_pairs = [
+            (target, positive_context)
+            for target, walk in zip(targets, walks)
+            for positive_context in walk[1:]
+        ]
+
+        negative_samples = self.random.choices(
+            all_nodes, weights=sampling_distribution, k=len(positive_pairs)
+        )
+        negative_pairs = [
+            (target, negative_context)
+            for (target, _), negative_context in zip(positive_pairs, negative_samples)
+        ]
+
+        labels = [1] * len(positive_pairs) + [0] * len(negative_pairs)
+
+        # zip and shuffle
+        edge_ids_labels = list(zip(positive_pairs + negative_pairs, labels))
+        self.random.shuffle(edge_ids_labels)
 
-                # (target,contect) pair sampling - GraphSAGE way
-                target = walk[0][0]
-                context_window = walk[0][1:]
-                for context in context_window:
-                    # Don't add self pairs
-                    if context == target:
-                        continue
-
-                    positive_pairs.append((target, context))
-                    sample_counter += 1
-
-                    # For each positive sample, add a negative sample.
-                    random_sample = self.random.choices(
-                        all_nodes, weights=sampling_distribution, k=1
-                    )
-                    negative_pairs.append((target, *random_sample))
-                    sample_counter += 1
-
-                    # If the batch_size number of samples are accumulated, yield.
-                    if sample_counter == batch_size:
-                        all_pairs = positive_pairs + negative_pairs
-                        all_targets = [1] * len(positive_pairs) + [0] * len(
-                            negative_pairs
-                        )
-
-                        positive_pairs.clear()
-                        negative_pairs.clear()
-                        sample_counter = 0
-
-                        edge_ids_labels = list(zip(all_pairs, all_targets))
-                        self.random.shuffle(edge_ids_labels)
-                        edge_ids, edge_labels = [
-                            [z[i] for z in edge_ids_labels] for i in (0, 1)
-                        ]
-
-                        yield edge_ids, edge_labels
+        # convert to batches
+        return [
+            tuple(zip(*edge_ids_labels[i : i + batch_size]))
```

Comment:
Hmm yes I think that could be possible 👍I'll give it a go

Reply:
Now it returns a tuple of numpy arrays, it seems like all the downstream usage still works

Reply:
Actually I might just revert this, since I realised that IDs are not necessarily integers, so then we'd need to have a numpy array of arbitrary python objects?

Reply:
I think that should work ~automatically after removing `dtype`s? It may make sense to manipulate an array of IDs an array of labels separately (I think that's why the final return has `[..., :2]`, `[..., 2]`, right?).

```python
>>> import numpy as np
>>> np.array(["foo", 1, [], {}])
array(['foo', 1, list([]), {}], dtype=object)
```

Reply:
Do you still suspect the "manipulations" will be faster though? 

The array of IDs and array of labels are zipped together for shuffling purposes - I guess we could just create another array of indices to shuffle instead, but it didn't occur to me that would necessarily be better

Reply:
Ah, hm... I don't know. It could be done with pandas or even a numpy `recarray` https://docs.scipy.org/doc/numpy/user/basics.rec.html 

```
>>> np.rec.array([("foo", "barbaz", 1), ("qux", "1234567890", 0)])
rec.array([('foo', 'barbaz', 1), ('qux', '1234567890', 0)],
          dtype=[('f0', '<U3'), ('f1', '<U10'), ('f2', '<i8')])
```

However, I suspect just shuffling the arrays "together" would be fine too, e.g. https://stackoverflow.com/q/4601373/1256624 (not the first answer, though, since that seems to rely on dtypes matching).